src/sidebar/util/thread.js

```diff
@@ -31,3 +32,61 @@ export function countHidden(thread) {
 export function countVisible(thread) {
   return countByVisibility(thread, true);
 }
+
+/**
+ * Find the topmost annotations in a thread.
+ *
+ * For the (vast) majority of threads, this is the single annotation at the
+ * top level of the thread hierarchy.
+ *
+ * However, when the top-level thread lacks
+ * an annotation, as is the case if that annotation has been deleted but still
+ * has replies, find the first level of descendants that has at least one
+ * annotation (reply) and return the set of annotations (replies) at that level.
+ *
+ * For example, given the (overly-complex) thread-annotation structure of:
+ *
+ * [missing]
+ *   - [missing]
+ *       - reply 1
+ *            - reply 2
+ *            - reply 3
+ *       - reply 4
+ *   - [missing]
+ *       - reply 5
+ *   - [missing]
+ *      - [missing]
+ *          - reply 6
+ *
+ * Return [reply 1, reply 4, reply 5]
+ *
+ * @param {Thread[]} threads
+ * @return {Annotation[]}
+ */
+export function rootAnnotations(threads) {
+  // If there are any threads at this level with extant annotations, return
+  // those annotations
+  const threadAnnotations = threads
+    .filter(thread => !!thread.annotation)
+    .map(thread => thread.annotation);
+
+  if (threadAnnotations.length) {
+    // TODO Argh :) typing
+    return /** @type {Annotation[]} */ (threadAnnotations);
```

Comment:
If you cast the result of the map function above (ie. `.map(thread => /** @type {Annotation} */ (thread.annotation))` or use your `notNull` helper). that will allow TS to infer `threadAnnotations` correctly. The issue is that TS doesn't understand the effect of `Array.filter`.